### ISSUE TITLE
logout method of Avatar is not called when RemoteReferences are GC'ed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7604]
|<img alt="syver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| syver reported|
|-|-|
|Trac ID|trac#392|
|Type|defect|
|Created|2003-11-10 22:39:23Z|
|Branch|https://github.com/twisted/twisted/tree/pb-avatar-decref-392|

Attachments:

* [patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/071/0715d58c74869d445849a688c0f3804892a5d6a1/d75b3b528276d7a9f30a04b8bccaf74e1d61f67a) (2761 bytes) - added by titty on 2004-03-12 18:58:25Z - 
* [patch2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/071/0715d58c74869d445849a688c0f3804892a5d6a1/9e26ce95028e83a061cf35e9b09b22c90d9b94e2) (2396 bytes) - added by titty on 2006-03-05 20:01:50Z - 
* [Test392.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/071/0715d58c74869d445849a688c0f3804892a5d6a1/0ec349bd100ad75a1a8e29d0028617d931e104ff.py) (890 bytes) - added by titty on 2006-03-05 20:02:38Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__392 392
type__defect defect
reporter__syver syver
priority__high high
milestone__ 
branch__branches_pb_avatar_decref_392 branches/pb-avatar-decref-392
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1068503963000000 1068503963000000
changetime__1469186923227445 1469186923227445
version__ 
owner__ 
cc__glyph cc__itamarst cc__warner cc__syver cc__titty cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="syver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|syver commented|
|-|-|

```
#!html
<pre>
the logout method that is returned from requestAvatar seems only to be called 
when disconnecting the broker (the tcp connection is broken). The 
RemoteReference class has a __del__ method that seems to decrease a serverside 
reference count but this has no effect on the Avatar. When having a long 
running twisted application that is a pb client and which requests different 
avatars this leads to unbounded numbers of allocated avatars.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I remember discussing this with glyph, yeah. I'm not sure what we decided, glyph?

What you should do is have a perspective_logout method that the remote side
calls, for one thing, instead of depending on gc. Although... yeah, that
wouldn't gc the avatar itself, just allow you to clean up some of its resources.
Guess we need some extra hooks.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Brian, you probably should look at the issue involved.

Another way of putting it might be "we need better handles for garbage
collection of RemoteReference objects". I.e. if we had a
"RemoteReference.youAreDead" method this bug would be easy to fix.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
Attached is a patch, which maintains weak refs on all RemoteReferences and
invalidates them, when connectionLost is called.I had the problem that as
soon as RemoteReferences had been instantiated, the garbage collector wouldn't
collect the broker.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

A unit test which failed without this patch would be awesome.

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

Attaching a cleaned up patch, and unittest.

```
ralf@breezy:~/Twisted$ bin/trial Test392.py
Running 2 tests.
Test392
  TestBrokerCollect
    test_gc_collect_broker ...                                           [FAIL]
  TestRemoteReferenceHash
    test RemoteReference.__hash__ with RemoteReferences created ...     [ERROR]

===============================================================================
[FAIL]: Test392.TestBrokerCollect.test_gc_collect_broker

  File "/home/ralf/Twisted/Test392.py", line 30, in test_gc_collect_broker
    self.failUnless(called, "broker alive and not garbage collected.")
twisted.trial.unittest.FailTest: broker alive and not garbage collected.
===============================================================================
[ERROR]: Test392.TestRemoteReferenceHash.test_hash

  File "/home/ralf/Twisted/Test392.py", line 15, in test_hash
    hash(ref)
exceptions.TypeError: __hash__() should return an int
-------------------------------------------------------------------------------
Ran 2 tests in 0.050s

FAILED (failures=1, errors=1)
ralf@breezy:~/Twisted$ patch -p0 <patch2 
patching file twisted/spread/pb.py
ralf@breezy:~/Twisted$ bin/trial Test392.py
Running 2 tests.
Test392
  TestBrokerCollect
    test_gc_collect_broker ...                                             [OK]
  TestRemoteReferenceHash
    test RemoteReference.__hash__ with RemoteReferences created ...        [OK]

-------------------------------------------------------------------------------
Ran 2 tests in 0.037s

PASSED (successes=2)
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [23930]) Branching to 'pb-avatar-decref-392'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

I'm not sure I understand the attached test and fix.  The ticket description is for avatar logout behavior; the attached code seems to be for breaking cycles after disconnection.

I fixed the issue in the ticket description in the branch.  Mayhaps another ticket should be opened for the other issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Ah, a very important thing that I definitely need to point out - `pb-avatar-decref-392` is relative to `pb-concurrent-logins-3295`.  The easiest thing for a review to do is to review [#3295](https://github.com/twisted/twisted/issues/3295) first and wait for it to be merged before trying to review this.

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|titty commented|
|-|-|

Sorry, I can't remember, why I attached my patch to this ticket. Anyway, this has been a real issue for us and we are running a locally modified version of twisted with this code...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

OK, the given branch seems to fix one of the problem. Please open a ticket with the other issue (we have a failing testcase, it should be easy to solve ;)), and merge this one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [23973]) Merge pb-avatar-decref-392

Author: exarkun
Reviewer: therve
Fixes: [#7604](https://github.com/twisted/twisted/issues/7604)

Change PB so that when a connection is lost or when the last
reference to an avatar is dropped, the logout method returned
by the realm is called.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|GitHub commented|
|-|-|

In changeset f25a60ad41ce5d8e07d800c2914278c952b1e84c
```
#!CommitTicketReference repository="" revision="f25a60ad41ce5d8e07d800c2914278c952b1e84c"
Merge pull request #392 from twisted/8653-einarfd-ssh-transport-service-name-2

Author: einarfd
Reviewer: rodrigc
Fixes: #8653

Change the ssh transport service name to a byte string
```
